src/relay/transforms/fp32_to_fp16.cc

```diff
@@ -0,0 +1,337 @@
+    Array<Expr> new_args;
+    Array<Type> new_arg_types;
+    for (size_t i = 0; i < cur_args.size(); i++) {
+      Expr cur_arg = cur_args[i];
+      Type cur_arg_type = cur_arg_types[i];
+      Expr new_arg = CastArg(cur_arg, cur_arg_type, wanted_dtype);
+      Type new_arg_type = GetType(new_arg);
+      new_args.push_back(new_arg);
+      new_arg_types.push_back(new_arg_type);
+    }
+    return {new_args, new_arg_types};
+  }
+
+ public:
+  explicit AmpGraphCreator(ColorFunc colorer, OutputDtypeFunc output_dtype_func)
+      : ExprMutator(), colorer(colorer), output_dtype_func(output_dtype_func) {}
+
+  Expr VisitExpr_(const CallNode* call_node) {
+    FP16ConversionCategory initial_color = colorer(call_node);
+    auto new_op = this->Mutate(call_node->op);
+
+    // Mutate arguments to FP16 form first if possible and keep track of whether all floating point
+    // tensors are in FP16 form already. This is useful for propagating color.
+    Array<Expr> new_args;
+    Array<Type> new_arg_types;
+    bool all_args_fp16_compatible = true;
+    for (Expr arg : call_node->args) {
+      Expr new_arg = this->Mutate(arg);
+      Type new_arg_type = GetType(new_arg);
+      new_args.push_back(new_arg);
+      new_arg_types.push_back(new_arg_type);
+
+      if (all_args_fp16_compatible) {
+        // We can cast Vars and Constants to the right types so don't care about the types.
+        bool is_fp16_compatible = IsFP16Type(new_arg_type, true) || arg->IsInstance<VarNode>() ||
+                                  arg->IsInstance<ConstantNode>();
+        all_args_fp16_compatible &= is_fp16_compatible;
+      }
+    }
+
+    // Determine the final color.
+    FP16ConversionCategory final_color;
+    if (initial_color == GRAY) {
+      final_color = all_args_fp16_compatible ? GREEN : RED;
+    } else {
+      final_color = initial_color;
+    }
+
+    // Create the new arguments to the call.
+    DataType wanted_arg_dtypes = final_color == GREEN ? DataType::Float(16) : DataType::Float(32);
+    auto call_args_and_types = CastAllArgs(new_args, new_arg_types, wanted_arg_dtypes);
+
+    Array<Expr> call_args = call_args_and_types.first;
+    Array<Type> call_arg_types;
+
+    if (call_node->op.as<FunctionNode>()) {
+      // Function Nodes don't store type info in the Call, it should be a []
+      call_arg_types = call_node->type_args;
+    } else {
+      call_arg_types = call_args_and_types.second;
+    }
+
+    // Finally create the new attributes.
+    if (final_color == GREEN) {
+      FP16OpDType output_dtypes = output_dtype_func(call_node);
+
+      Attrs new_attrs = GetNewAttrs(call_node, output_dtypes.accumulation_dtype);
+      Expr output = Call(new_op, call_args, new_attrs, call_arg_types, call_node->span);
+      if (output_dtypes.accumulation_dtype != output_dtypes.output_dtype) {
+        output = CastArg(output, GetType(output), output_dtypes.output_dtype);
```

Comment:
Wouldn't this introduce unnecessary cast ops? For example, the accumulation dtype is FP32 and the followed op is RED. Will this make it `A(GREEN) - cast_to_fp16 - cast_to_fp32 - B(RED)`?

Reply:
Hmm I don't believe so since `CachedCast` will also cache the reverse result.

E.g. `CachedCast(A(Green), FP16)` would produce `A(GREEN) - cast_to_fp16` 

But internally it would cache:
`Node, wanted_dtype`
`A(GREEN), FP16` --> `cast_to_fp16`
`cast_to_fp16, FP32` --> `A(GREEN)`

So attempting to cast `cast_to_fp16` to `fp32` would return `A(GREEN)`

It would be worth having a test case to cover this however and make sure.

Reply:
I see...this mechanism is interesting and I haven't paid too much attention on it. At the first glance, I would worry if the cache will blow up when the model is too large, but I'll probably take a deeper look at this mechanism later.